### PR TITLE
Extract documentaion_url from pypi

### PR DESCRIPTION
**Pull request by FarhanAliRaza**
Fixes https://github.com/djangopackages/djangopackages/issues/1179

### CODE REVIEW

File: package/models.py [hunk 1]
```diff
@@ -375,6 +380,24 @@ def fetch_pypi_data(self):
                 if self.pypi_downloads is None:
                     self.pypi_downloads = total_downloads
 
+                # get documents_url from pypi
+                if self.documentation_url is None:
+                    if info["project_urls"].get("Documentation"):
```
Comment (saadmk11) [inline package/models.py:385]: After looking at the PyPI API, it seems like the name of this field (`"Documentation"`) can be anything. So, I think we can try all the common variants as well. e.g: `docs`

File: package/utils.py [hunk 2]
```diff
@@ -72,3 +72,13 @@ def normalize_license(license: str):
             return "Custom"
         return stripped_license
     return "UNKNOWN"
+
+
+def extract_documentation_url_from_markdown(description):
```
Comment (saadmk11) [inline package/utils.py:77]: Calling this function with `description="[Documentation](This is not a URL)"` will return an invalid URL. If we **really** need to extract the documentation URL from the `description` perhaps we could use some kind of `regex`? 

**Note:** A really large description filed might cause some performance issues.

File: package/models.py [hunk 3]
```diff
@@ -375,6 +380,24 @@ def fetch_pypi_data(self):
...
+                        self.documentation_url = info["project_urls"].get(
+                            "Documentation"
+                        )
+                    elif info.get("docs_url"):
```
Comment (saadmk11) [inline package/models.py:389]: Could we check this field first?

**Review by saadmk11 — COMMENTED** [review]
Hi @FarhanAliRaza, excellent work on this. :100: 

I just added some small nitpicks and suggestions. :)

File: package/models.py (hunk 3 above)
Comment (FarhanAliRaza) [reply package/models.py:389]: I have seen that it is null for all the packages, I have seen but the name suggests it should contain the docs URL. I could not find any documentation for fields of JSON API. Can you share if you have any?. We can remove it if you think.

File: package/utils.py (hunk 2 above)
Comment (FarhanAliRaza) [reply package/utils.py:77]: Django has its own `URLValidator` Can we use that? URL regex gets pretty complex.

Comment (FarhanAliRaza) [reply package/utils.py:79]: I did a simple test. How can I improve it?
image

File: package/models.py [hunk 4]
```diff
@@ -375,6 +380,32 @@ def fetch_pypi_data(self):
...
+                        self.documentation_url = docs_url
+                    elif (
+                        info.get("description")
+                        and info.get("description_content_type") == "text/markdown"
+                    ):
+                        self.documentation_url = (
+                            extract_documentation_url_from_markdown(
+                                info.get("description")
+                            )
+                        )
```
Comment (jefftriplett) [inline package/models.py:407]: I think we can limit the scope to only what is structured data from PyPI JSON API. 

We have another tool we worked on which maintainers can use to help them score better that we still need to hook up.

File: package/models.py (hunk 3 above)
Comment (saadmk11) [reply package/models.py]: Interesting, Let's keep it for now.

File: package/models.py (hunk 4 above)
Comment (saadmk11) [reply package/models.py:407]: I agree.

Comment (FarhanAliRaza) [reply package/models.py:407]: removed

**Review by jefftriplett — APPROVED** [review]

**Comment by jefftriplett** [comment]
@FarhanAliRaza good work! I tested it locally and it was hitting the correct values. I'll push a test, and then going forward, we can go over tests. 

The only thing that's missing would be the nightly pypi updater knowing to hit the right endpoints. I never speced that out fully, but I can.

**Pull request merged**